### ISSUE TITLE
Seed Swap ∘ Split_reduce: conv-gradient accumulations are rejected for inner accumulation symbols, so the dominant reduction is unreachable

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#537]
Successor to #484, whose four tasks are complete but whose payoff is unrealized. `Split_reduce` is implemented, seeded, and times cleanly (42 candidates timed across 7 searches on CUDA, 0 failed, 0 dedup'd) — but it is never proposed for the reductions it was filed for.

## The rejection, pinned

Not the published `sr_out_max` / `sr_red_min` gates — those pass. It is `Sched.op_legality`, and it rejects **every** axis of the conv-gradient accumulations. From the probe added in lukstafi/ocannl-staging#255:

```
w:bias_conv1.grad(6) loops[i527=64s,i528=28s,i529=28s,i530=6s]
    axis i527 extent 64 -> illegal: the accumulation cell mentions i530, which is not bound
                                    by a loop enclosing the reduction loop in this statement
    axis i530 extent 6  -> illegal: the accumulation cell mentions the reduction loop i530
```

`Split_reduce` v1 requires the accumulation cell's symbols to be bound *outside* the reduction loop. OCANNL lowers conv bias gradients with the output-channel loop **innermost** — the exact inverse. Same verdict on every parameter-gradient accumulation in the network.

The recognizer's own message already names the missing prerequisite: *"Swap it inside `<axis>` (or split-reduce it) first."* The composition is expressible; seeding simply never proposes it.

## The task

Seed `Swap` ∘ `Split_reduce` — reorder the accumulation loop outside the reduction axis, then split. Concretely:

1. When `op_legality` rejects a split-reduce site for the inner-accumulation-symbol reason specifically, attempt the enabling `Swap` rather than dropping the site. The verdict already distinguishes this cause from other rejections, so the recognizer can be made to emit the composition instead of a decline.
2. Verify the `Swap` is itself legal for the site (it is a reduction-order change, which `op_legality` licenses under the reassociation rules for `Swap`'s accumulation contract — worth confirming per-site rather than assuming).
3. Pin the composition with a test in the `autotune_split_reduce` family, asserting `split_reduce_timed > 0` on a conv-gradient-shaped site — not merely `split_reduce_candidates`.

## Why this is the whole remaining payoff

After gh-527's fix, HIP's lenet segment fell from 350.712 ms to 72.76 ms (4.8x) — that portion was the product-space gate, not the reduction. The residual **~73 ms is the inherent serial reduction**, still **89.3%** of the step (cifar_stride: 78.7%). The HIP census names the shape: a serial nest over `[64,28,28,6]` producing 6 cells on one thread.

Three backends measured this same segment as dominant (Metal 69-91%, CUDA 91-95%, HIP 89-97% post-gh-527). Nothing else in the schedule space currently addresses it: #521's fixes concern tensorized candidates, and #500's blocking work was deprioritized by exactly this attribution.

## Verification

End-to-end on the machines where the segment was measured — HIP and CUDA — via `BENCH_SEG_TIMES=1` on `bench_conv_diag`, against the post-gh-527 numbers above rather than the older `bd075cd5` tables. Note lukstafi/ocannl-staging#255's caution: pre-gh-521 conclusions about the sketch families should be discarded, since before that fix every `F_sketch[gpu ...]` failed to compile on HIP and split-reduce won `mlp_small` by default.

**Comment by lukstafi** [comment]
## CUDA leg: confirmed, and the interchange removes ~90% of the segment

Paired same-session A/B on the RTX 5070 Ti Laptop / CUDA 13.3 box (the machine
[report-gh484-cuda.md](https://github.com/lukstafi/ocannl-staging/blob/master/benchmarks/report-gh484-cuda.md)
came from). `OLD` = staging master `081e70e2` with gh-537's `` `Hoist `` re-probe branch disabled,
so `splittable` returns exactly its pre-gh-537 verdicts; the library delta is that one branch, and
the `NEW` binary was rebuilt after restoring the source and is byte-identical to the one measured.
Two replicates per side per workload, interleaved in one session. Full writeup:
[`benchmarks/report-gh537-cuda.md`](https://github.com/lukstafi/ocannl-staging/blob/master/benchmarks/report-gh537-cuda.md).

**Not a null result.** The cleanest evidence needs no cross-run comparison — one `NEW` lenet
search, arm A, same process, same code, only the schedule differing between lines:

```
F_preset[bs=cfg cfg-thresh]:                     114.7615 ms
F_preset[bs=64 priv]:                            114.4498 ms
F_split[cross_entropy red64 out1 b32]:           114.7957 ms
F_split[bias_conv1.grad red64 out6 b32 swap3]:    13.8552 ms   <---
F_split[b_logits.grad red64 out10 b32 swap1]:    114.3252 ms
F_split[bias_conv2.grad red64 out16 b32 swap3]:  113.2314 ms
F_split[all four sites]:                          12.2449 ms
untuned-default in-process control:              114.5708 ms
```

Splitting **one** site — `bias_conv1.grad`, the write of the `loops[64s,28s,28s,6s]` → 6 cells nest
this issue names, reached via the 3-swap interchange — is **8.3x**. Every classifier-head site
gh-484 already reached moves nothing, exactly as the predecessor leg found.

`BENCH_SEG_TIMES=1` sizes what was removed. Default placements, min of 20:

| | top segment | ms | share | tuned arm-A result |
|---|---|---|---|---|
| lenet | seg22 `bias_conv1.grad n65.grad` | 102.47 | 87.5% of 117.06 | 114.7 → **12.2 ms** |
| cifar_conv | seg22 `bias_conv1.grad n65.grad` | 426.36 | 82.1% of 519.53 | 522.5 → **122.3 ms** |

The totals match the in-process untuned controls to within noise, so this is a like-for-like
accounting: **~90% of the segment on lenet, ~94% on cifar_conv, gone.**

### Detection

`BENCH_SR_SITES=1`, all three workloads: 2 sites → **4**, with `bias_conv1.grad` present on each and
tagged with its interchange (`via 3 swaps: i530^i529 i530^i528 i530^i527` on lenet). Per lenet
search `split_reduce_timed` goes 6 → **10** (5 per `tune_placements` arm), **0 FAILED, 0 dedup'd**,
and the crowned schedule is `F_split_saved` containing the conv-gradient site — not merely seeded.
Task 3's `split_reduce_timed > 0` on a conv-gradient-shaped site is satisfied end to end, not only
in the unit test.

### Where it does and does not reach the user

`tune_placements` ships the materialize-all arm on every conv workload, and that arm already routes
around the serial nest by a different mechanism (untuned 11.7 ms vs arm A's 114.6 ms on lenet). In
that arm the split wins only where the site is narrow. Replayed tuned artifact, p50 of 100 synced
steps, two replicates:

| workload | OLD | NEW | |
|---|---|---|---|
| lenet | 11.589 / 11.583 ms | **10.794 / 10.790 ms** | **−6.9%** |
| cifar_stride | 33.06 / 31.19 ms | 31.17 / 31.20 ms | neutral |
| cifar_conv | 81.54 / 81.64 ms | 81.73 / 81.61 ms | neutral |

Within-side spread is 0.06% on lenet, so that delta is ~100x its noise. The cifar workloads' site
has 32 output cells instead of 6; there the split costs 60% (`F_split[bias_conv1.grad out32 swap3]`
131.7 ms vs the best preset's 81.8 ms), the tuner measures that and declines it. **The mechanism is
safe where it does not help** — the neutral cells are "no gain", not "harm", at a cost of two extra
candidates per arm. Search cost is unchanged within run-to-run variation; cifar losses are
bit-identical across the pair and lenet differs by ≤2.4e-07 (only lenet crowns a different
schedule).

### Two things this leg turned up

**1. `sr_max_sites = 4` is now binding, and it excludes every weight gradient.** The probe tags
`kernel_conv1.grad` (150 cells), `kernel_conv2.grad` (2400), `w_logits.grad`, `b_fc1.grad`,
`b_fc2.grad` all `[hoistable: …]` — the interchange would unblock them — but `split_reduce_sites`
ranks by `sr_red / sr_out` and truncates at 4. Every weight gradient has more cells than its
reduction extent, so integer division sends each to ratio 0, and the three biases plus the loss take
all four slots. Before gh-537 the cap never bound (only 2 sites were found), so this is a new
question rather than a regression.

That matters because of:

**2. The dominant segment of the arm that ships is a different one.** `BENCH_MATERIALIZE=1
BENCH_SEG_TIMES=1`:

| workload | top segment | ms | share |
|---|---|---|---|
| lenet | seg33 `kernel_conv1.grad bias_conv1.grad` (+ SGD update) | 7.39 | 59.5% of 12.41 |
| cifar_conv | seg33 `kernel_conv1.grad bias_conv1.grad` | 49.24 | 49.3% of 99.96 |

Both fuse the **kernel** gradient with the bias gradient. gh-537 splits only the bias half — worth
7% on lenet where 6 cells make it profitable, nothing on cifar_conv where the kernel half dominates.
So the obvious follow-up is seeding the weight gradients: raise the cap, or rank by estimated
segment cost rather than `sr_red / sr_out`. This leg cannot say whether they would time well — they
were never seeded, so they were never timed, and that is exactly the seeded-vs-timed distinction
this arc has been bitten by twice.

### On the stated verification instrument

This issue asks for verification via `BENCH_SEG_TIMES=1`. That instrument measures the **default
pipeline**, which an autotune-only change cannot affect — the gh-484 leg made the same observation.
It is the right instrument for locating the segment and for sizing what a split has to remove (both
tables above use it), but the one that answers "did it get faster" is the tuned candidate timing,
where the split is actually applied.

The Metal leg of this measurement is running separately; HIP has not been re-measured here.

**Comment by lukstafi** [comment]
## Measured on Metal: yes, decisively in arm A; −6.6% where it ships on lenet

Full writeup: `benchmarks/report-gh537-metal.md` (staging PR #263). Companion to the CUDA leg (staging #261, `report-gh537-cuda.md`) — **the two backends agree**.

Measured at `ad46ab21`, after ahrefs/ocannl#532's fix landed. That was necessary, not cautious: every search here logs `baseline: NOT DISPATCHED, binds no hardware dimension on metal`, and #532 measured 6.9 s/run for an unparallelized lenet baseline on Metal against a 35.7 ms winner. A measurement spanning that fix would have compared different candidate pools.

**A/B**: `OLD` disables only the `` `Hoist `` re-probe this issue's fix added to `Autotune.split_reduce_sites`; `NEW` rebuilt byte-identical after restoring the source. Interleaved, one session, two replicates per side, fresh cache per search, idle machine.

| workload | arm A (default placements) | arm B (ships) | shipped p50 |
|---|---|---|---|
| lenet | 249.1 → **39.9 ms** (6.3x) | 35.4 → **33.2 ms** | 35.59 → **33.22 ms** (−6.6%) |
| cifar_stride | 240.7 → **122.3 ms** (2.0x) | 95.7 → 95.3 ms | neutral |
| cifar_conv | 1270.1 → **369.2 ms** (3.4x) | 276.1 → 275.4 ms | neutral |

One site accounts for all of it. In a single `NEW` lenet process, arm A — every preset and every *other* split site sits at 249–251 ms:

```
F_preset[bs=64 priv]:                             249.10 ms
F_split[cross_entropy red64 out1 b32]:            250.59 ms
F_split[bias_conv1.grad red64 out6 b32 swap3]:     44.40 ms   <---
F_split[b_logits.grad red64 out10 b32 swap1]:     250.45 ms
F_split[bias_conv2.grad red64 out16 b32 swap3]:   249.52 ms
F_split[all four sites]:                           39.86 ms
untuned-default in-process control:               250.42 ms
```

Sites 2 → 4; crowned schedule `F_saved` → `F_split_saved` in **both** arms. `split_reduce` candidates reach timing rather than being seeded-and-dropped — the failure mode this arc hit twice does not recur here.

Sizing, `BENCH_SEG_TIMES=1`, default placement: seg22 `bias_conv1.grad n65.grad` is 214.49 of 256.41 ms on lenet (83.6%), 956.25 of 1305.00 on cifar_conv (73.3%), 139.61 of 258.30 on cifar_stride (54.1%). Against arm A, the fix removes ~98% / ~94% / ~84% of it.

**Where it ships, profitability tracks the site's output width** — 6 cells −6.6% (≈40x the within-side spread), 32 cells a loss the tuner measures and declines. The cifar cells are "no gain", not "harm".

### A note on this issue's stated instrument

#537 asks for verification via `BENCH_SEG_TIMES=1` on `bench_conv_diag`. That measures the *default* pipeline, which an autotune-only change cannot move — it is right for locating the segment and sizing what a split must remove, but the tuned candidate timing is what answers "did it get faster". Same observation the CUDA leg made independently.

Also worth recording: the per-segment instrument over-counts the fused step by +2.4% on the default placement but **+21.9%** on the materialized one, whose segments are more numerous and smaller. Shares within a placement are usable; a seg-time total is not a step time.

**Comment by lukstafi** [comment]
Implemented and measured. The payoff is real but **placement-dependent**, which is the finding worth carrying forward.

**Implementation** (lukstafi/ocannl-staging#257): `Schedule.split_reduce_hoist` reports which accumulation-cell symbols are bound inside the reduction loop — structurally, from the recognizer's own hermetic probe rather than parsed from messages — and `split_reduce_sites` builds an interchange chain hoisting each one outward, confirms every `Swap` is `Op_legal` against the code it acts on, and re-probes the split on the interchanged code. The chain is recorded on the site and replayed by the `F_split` candidate's prelude.

**Measured on CUDA** (`benchmarks/report-gh537-cuda.md`, lukstafi/ocannl-staging#261; paired same-session A/B with the `Hoist` re-probe branch disabled for the OLD side, two replicates each, interleaved):

- **Not a null result.** In one lenet search, one process, only the schedule differing: every preset and every other split site lands at 113-115 ms, while `F_split[bias_conv1.grad red64 out6 b32 swap3]` lands at **13.86 ms — 8.3x**. `BENCH_SEG_TIMES` sizes what was removed: seg22 was 102.47 of 117.06 ms on lenet and 426.36 of 519.53 ms on cifar_conv. **~90% of that segment, gone.**
- Sites 2 → 4, `split_reduce_timed` 6 → 10 per lenet search, 0 FAILED, 0 dedup'd, and the crowned schedule is `F_split_saved` containing the conv-gradient site.

**The shipping artifact moves much less, and this corrects a framing this cluster carried since the sweep.** `tune_placements` ships the materialize-all arm, which *already* routes around the serial nest. lenet's site has 6 output cells: 11.586 → 10.792 ms, **−6.9%** against 0.06% within-side spread. The cifar workloads' site has 32 cells, the split costs 60%, and the tuner measures that and correctly declines it — both cells neutral, no gain and no harm.

So the 94.6% / 91.4% / 89.3% segment shares that drove this issue's priority were of the **default** pipeline. The tuned artifact ships a different placement whose dominant segment is `seg33` (`kernel_conv1.grad` + `bias_conv1.grad`) at 59.5% / 49.3%. A segment share needs a placement qualifier; this cluster's earlier framing did not have one.

**Successor: #541.** `sr_max_sites = 4` is now binding — before this change only 2 sites were ever found, so the cap never bound — and the `sr_red / sr_out` ranking sends every weight gradient to ratio 0 by integer division, so the three biases plus the loss take all four slots. `kernel_conv1.grad` is the untouched half of the shipping arm's dominant segment, and the obvious next experiment.
